### PR TITLE
virtiofs: Run unprivileged without feature gate

### PR DESCRIPTION
**Pull request by germag**
**What this PR does / why we need it**:

Currently, virtiofsd can only run unprivileged for sharing ConfigMaps, Secrets, etc., but to be able to share filesystem PVs requires the feature gate `ExperimentalVirtiofsSupport`.

This PR removes that restriction, making it possible to share filesystem PVs using unprivileged virtiofsd if the feature gate is disabled, and continue running as root if the feature gate is enabled. So this change doesn't affect users that are already running a privileged virtiofsd.

Running virtiofsd without privileges has some limitations, for instance, virtiofsd cannot switch uids/gids so all the new files created inside the PV will be owned by the same uid/gid (the virtiofsd uid/gid) regardless of the uid/gid of the process that created them inside the VM.  After the merge, all the limitations and workarounds will be documented (see https://github.com/kubevirt/user-guide/pull/734).

Signed-off-by: German Maglione <gmaglione@redhat.com>
Tested-by: Javier Cano Cano <jcanocan@redhat.com>

**Release note**:

```release-note
Exposing Filesystem Persistent Volumes (PVs)  to the VM using unprivilege virtiofsd.
```

### CODE REVIEW

File: pkg/virt-controller/services/virtiofs_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+		container := generateVirtioFSContainers(vmi, "virtiofs-container", config)
+		Expect(container).To(HaveLen(2))
+
+		if shouldEnableFeatureGate {
+			// PV
+			Expect(container[0].SecurityContext.RunAsNonRoot).To(HaveValue(BeFalse()))
+			Expect(container[0].SecurityContext.AllowPrivilegeEscalation).To(HaveValue(BeTrue()))
+			// Secret
+			Expect(container[1].SecurityContext.RunAsNonRoot).To(HaveValue(BeTrue()))
```
Comment (0xFelix) [inline pkg/virt-controller/services/virtiofs_test.go:87]: Shouldn't this match with `BeFalse()`?

File: pkg/virtiofs/virtiofs.go [hunk 2]
```diff
@@ -18,10 +19,10 @@ func VirtioFSSocketPath(volumeName string) string {
 	return filepath.Join(VirtioFSContainersMountBaseDir, socketName)
 }
 
-// RequiresRootPrivileges Returns true if the volume requires the virtiofs
+// CanRunWithPrivileges Returns true if the volume requires the virtiofs
 // container to run as user root
```
Comment (0xFelix) [inline pkg/virtiofs/virtiofs.go:23]:
```suggestion
// CanRunWithPrivileges Returns true if the virtiofs container of the volume 
// can run as user root
```

File: tests/virtiofs/datavolume.go [hunk 3]
```diff
@@ -75,25 +75,21 @@ var _ = Describe("[sig-storage] virtiofs", decorators.SigStorage, func() {
...
-
-		AfterEach(func() {
-			libstorage.DeletePVC(pvc1, testsuite.NamespacePrivileged)
-			libstorage.DeletePVC(pvc2, testsuite.NamespacePrivileged)
-		})
-
-		DescribeTable("should be successfully started and accessible", func(option1, option2 libvmi.Option) {
+		DescribeTable("should be successfully started and accessible", func(namespace string, option1, option2 libvmi.Option) {
+			if namespace == testsuite.NamespacePrivileged {
+				checks.SkipTestIfNoFeatureGate(virtconfig.VirtIOFSGate)
```
Comment (0xFelix) [inline tests/virtiofs/datavolume.go:87]: Who enables the feature gate? Do we need this skip?

File: tests/virtiofs/datavolume.go [hunk 4]
```diff
@@ -145,190 +141,191 @@ var _ = Describe("[sig-storage] virtiofs", decorators.SigStorage, func() {
...
+			checks.SkipTestIfNoFeatureGate(virtconfig.VirtIOFSGate)
+		}
 
-			dataVolume = libdv.NewDataVolume(
-				libdv.WithRegistryURLSource(cd.DataVolumeImportUrlForContainerDisk(cd.ContainerDiskAlpine)),
-				libdv.WithPVC(libdv.PVCWithStorageClass(sc)),
-			)
-		})
+		if !libstorage.HasCDI() {
+			Skip("Skip DataVolume tests when CDI is not present")
```
Comment (0xFelix) [inline tests/virtiofs/datavolume.go:263]: Are the tests expected to run without CDI at all?

File: pkg/virt-controller/services/virtiofs_test.go (hunk 1 above)
Comment (germag) [reply pkg/virt-controller/services/virtiofs_test.go:87]: Nop, for a Secret, ConfigMap, DownwardApi, and ServiceAccount, virtiofsd will always run as non-root, even if the feature gate is enabled

File: tests/virtiofs/datavolume.go (hunk 4 above)
Comment (germag) [reply tests/virtiofs/datavolume.go]: No, we skip the test if CDI is not present (that check was already in the test before my changes)

**Comment by alicefr** [comment]
The change looks good to me
/lgtm
/cc @xpivarc

File: pkg/virt-controller/services/virtiofs_test.go (hunk 1 above)
Comment (0xFelix) [reply pkg/virt-controller/services/virtiofs_test.go:87]: ACK

File: tests/virtiofs/datavolume.go (hunk 4 above)
Comment (0xFelix) [reply tests/virtiofs/datavolume.go]: Yes, my question was more aimed at whether KubeVirt tests run at all without CDI. IOW can we expected CDI always to be present when the tests run (without the need for another check or skip, not present is expected to result in failure)? IIRC it is present by default in kubevirtci.

Comment (germag) [reply tests/virtiofs/datavolume.go]: Ohh, I don't know really, maybe this was copy-pasted from `tests/storage/datavolume.go`, that was the original location of this test  (btw, that test checks twice per test if the CDI is present).

I can remove the check if is unnecessary

Comment (germag) [reply tests/virtiofs/datavolume.go]: Are these tests only intended for this CI or also for someone who does not have/needs CDI?

Comment (0xFelix) [reply tests/virtiofs/datavolume.go]: I can only speak for CNV where CDI is present in downstream tests.

**Comment by alicefr** [comment]
IMO, the code looks good
/approve
/hold
Wait for another round of reviews, @0xFelix @xpivarc, does this change look sane to you?

**Review by 0xFelix — APPROVED** [review]
/lgtm

Looks good to me, leaving hold in place to give @xpivarc a chance to look at it.

File: tests/virtiofs/datavolume.go [hunk 5]
```diff
@@ -251,34 +287,55 @@ var _ = Describe("[sig-storage] virtiofs", decorators.SigStorage, func() {
...
+		)
 	})
 
 	Context("Run a VMI with VirtIO-FS and a datavolume", func() {
-		var dataVolume *cdiv1.DataVolume
+		var (
+			dataVolume        *cdiv1.DataVolume
+			isVirtioFSEnabled bool
+			sc                string
+			exists            bool
```
Comment (xpivarc) [inline tests/virtiofs/datavolume.go:301]: `exists` could be local

File: tests/virtiofs/datavolume.go [hunk 6]
```diff
@@ -251,34 +287,55 @@ var _ = Describe("[sig-storage] virtiofs", decorators.SigStorage, func() {
...
-
-			dataVolume = libdv.NewDataVolume(
-				libdv.WithRegistryURLSource(cd.DataVolumeImportUrlForContainerDisk(cd.ContainerDiskAlpine)),
-				libdv.WithPVC(libdv.PVCWithStorageClass(sc)),
-			)
 		})
 
 		AfterEach(func() {
 			libstorage.DeleteDataVolume(&dataVolume)
+			if isVirtioFSEnabled {
```
Comment (xpivarc) [inline tests/virtiofs/datavolume.go:318]: This is redundant, the testsuite reverts the KV. This is rendering the `isVirtioFSEnabled` redundant.

File: tests/virtiofs/datavolume.go [hunk 7]
```diff
@@ -251,34 +287,55 @@ var _ = Describe("[sig-storage] virtiofs", decorators.SigStorage, func() {
...
+		},
+			Entry("unprivileged virtiofsd", util.NamespaceTestDefault),
+			Entry("privileged virtiofsd", testsuite.NamespacePrivileged),
+		)
 	})
 
 	Context("Run a VMI with VirtIO-FS and a datavolume", func() {
-		var dataVolume *cdiv1.DataVolume
+		var (
+			dataVolume        *cdiv1.DataVolume
```
Comment (xpivarc) [inline tests/virtiofs/datavolume.go:298]: `dataVolume` can be local

**Review by xpivarc — COMMENTED** [review]
I am a little bit worried that this will give the wrong impression that you could share any volume. Can we clarify this in the PR description + release note?

File: pkg/virt-controller/services/virtiofs_test.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+			},
+		},
+		Status: v1.KubeVirtStatus{
+			Phase:               v1.KubeVirtPhaseDeploying,
+			DefaultArchitecture: "amd64",
+		},
+	}
+	config, _, kvInformer := testutils.NewFakeClusterConfigUsingKV(kv)
+
+	enableFeatureGate := func(featureGate string) {
```
[the remark this answers is not in the document]
Comment (germag) [reply pkg/virt-controller/services/virtiofs_test.go:34]: done

File: tests/virtiofs/datavolume.go [hunk 9]
```diff
@@ -75,25 +75,38 @@ var _ = Describe("[sig-storage] virtiofs", decorators.SigStorage, func() {
...
-			checks.SkipTestIfNoFeatureGate(virtconfig.VirtIOFSGate)
-			createPVC(pvc1)
-			createPVC(pvc2)
+			isVirtioFSEnabled = checks.HasFeature(virtconfig.VirtIOFSGate)
 		})
 
 		AfterEach(func() {
-			libstorage.DeletePVC(pvc1, testsuite.NamespacePrivileged)
-			libstorage.DeletePVC(pvc2, testsuite.NamespacePrivileged)
+			if isVirtioFSEnabled {
```
[the remark this answers is not in the document]
Comment (germag) [reply tests/virtiofs/datavolume.go]: done

File: tests/virtiofs/datavolume.go (hunk 5 above)
Comment (germag) [reply tests/virtiofs/datavolume.go]: done

File: tests/virtiofs/datavolume.go (hunk 6 above)
Comment (germag) [reply tests/virtiofs/datavolume.go]: done

File: tests/virtiofs/datavolume.go (hunk 7 above)
Comment (germag) [reply tests/virtiofs/datavolume.go]: done

**Comment by germag** [comment]
> I am a little bit worried that this will give the wrong impression that you could share any volume. Can we clarify this in the PR description + release note?

Do you mean block PVs?, if so, is it enough calling them "filesystem PVs"?
 could you expand on which ones cannot be shared using virtiofs?

**Comment by xpivarc** [comment]
> > I am a little bit worried that this will give the wrong impression that you could share any volume. Can we clarify this in the PR description + release note?
> 
> Do you mean block PVs?, if so, is it enough calling them "filesystem PVs"? could you expand on which ones cannot be shared using virtiofs?

Yes, the filesystem PVs sound good. What I am more worried about is the privileges issue, users need to be aware there is a limitation to it.

**Comment by germag** [comment]
> > > I am a little bit worried that this will give the wrong impression that you could share any volume. Can we clarify this in the PR description + release note?
> > 
> > 
> > Do you mean block PVs?, if so, is it enough calling them "filesystem PVs"? could you expand on which ones cannot be shared using virtiofs?
> 
> Yes, the filesystem PVs sound good. What I am more worried about is the privileges issue, users need to be aware there is a limitation to it.

We plan to update the documentation, with all the limitations, as soon this PR is merged

**Comment by xpivarc** [comment]
> > > > I am a little bit worried that this will give the wrong impression that you could share any volume. Can we clarify this in the PR description + release note?
> > > 
> > > 
> > > Do you mean block PVs?, if so, is it enough calling them "filesystem PVs"? could you expand on which ones cannot be shared using virtiofs?
> > 
> > 
> > Yes, the filesystem PVs sound good. What I am more worried about is the privileges issue, users need to be aware there is a limitation to it.
> 
> We plan to update the documentation, with all the limitations, as soon this PR is merged

Would you mind posting placeholder PR?

**Pull request merged**